deepwell/src/services/site_member/service.rs

```diff
@@ -0,0 +1,158 @@
+
+        // Insert new membership.
+        let model = site_member::ActiveModel {
+            user_id: Set(user_id),
+            site_id: Set(site_id),
+            deleted_at: Set(None),
+            ..Default::default()
+        };
+
+        let membership = model.insert(txn).await?;
+
+        Ok(Some(membership))
+    }
+
+    /// Remove a user from a site.
+    pub async fn remove(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<SiteMemberModel> {
+        let txn = ctx.transaction();
+        tide::log::info!(
+            "Removing the membership of user ID {user_id} from site ID {site_id}"
+        );
+
+        let model = match Self::get_optional(ctx, SiteMembership { site_id, user_id })
+            .await?
+        {
+            // If membership is found, remove it by setting the leave date.
+            Some(member_model) => {
+                let mut model = member_model.into_active_model();
+                model.deleted_at = Set(Some(now()));
+                model.update(txn).await?
+            }
+
+            // If no membership is found, return BadRequest error.
+            None => {
+                tide::log::error!(
+                    "Could not remove user ID {user_id} from site ID {site_id} as they are not a member."
+                );
+                return Err(Error::BadRequest);
+            }
+        };
+
+        Ok(model)
+    }
+
+    #[inline]
+    pub async fn get(
+        ctx: &ServiceContext<'_>,
+        key: SiteMembership,
+    ) -> Result<SiteMemberModel> {
+        find_or_error(Self::get_optional(ctx, key)).await
+    }
+
+    /// Get whether a user is a member of a site or not. Returns `None` if no membership is found.
+    pub async fn get_optional(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<Option<SiteMemberModel>> {
+        let txn = ctx.transaction();
+        let model = SiteMember::find()
+            .filter(
+                Condition::all()
+                    .add(site_member::Column::SiteId.eq(site_id))
+                    .add(site_member::Column::UserId.eq(user_id))
+                    .add(site_member::Column::DeletedAt.is_null()),
+            )
+            .one(txn)
+            .await?;
+        Ok(model)
+    }
+
+    /// Get all users of a site, ordered by oldest to newest.
+    pub async fn get_site_members(
+        ctx: &ServiceContext<'_>,
+        site_id: i64,
+    ) -> Result<Vec<SiteMemberModel>> {
+        let txn = ctx.transaction();
+
+        let models = SiteMember::find()
+            .filter(
+                Condition::all()
+                    .add(site_member::Column::SiteId.eq(site_id))
+                    .add(site_member::Column::DeletedAt.is_null()),
+            )
+            .order_by_asc(site_member::Column::MembershipId)
+            .all(txn)
+            .await?;
+
+        Ok(models)
```

Comment:
This method needs a TODO stub for pagination. Sites like EN have tons of members, returning them all in one request doesn't make sense.

Reply:
Sure, I'll add a TODO